### ISSUE TITLE
Potential bug in the Product hook CRUD methods

### ISSUE THREAD

**Issue by jouino16** [PrestaShop/PrestaShop#28322]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

While implementing product hooks in a module and reading the code, I noticed a call to both save and update hooks in the `Product::update` method as follows:

```
Hook::exec('actionProductSave', ['id_product' => (int) $this->id, 'product' => $this]);
Hook::exec('actionProductUpdate', ['id_product' => (int) $this->id, 'product' => $this]);
```

Knowing that in the `Product::add` method, there is already a call to the hook:
```
Hook::exec('actionProductSave', ['id_product' => (int) $this->id, 'product' => $this]);
```

Isn't it a bug?

Personally, I avoided implementing those 2, and I have implemented `hookActionObjectProductAddAfter` and `hookActionObjectProductUpdateAfter` instead, as I want my hook to be called once.

### Expected behavior

The `Product::update` method should call the following hook:

```
Hook::exec('actionProductUpdate', ['id_product' => (int) $this->id, 'product' => $this]);
```

### Steps to reproduce

Check the `/classes/Product.php` file

### PrestaShop version(s) where the bug happened

1.7.8.5

**Comment by marwachelly** [comment]
Hello @jouino16,

Could you please provide us with clear steps to reproduce the issue? Clear description , screen shot , screen record..
You don't know how to get this information? Please read the following article: [How to create a bug report](http://build.prestashop.com/howtos/misc/how-to-create-bug-report)

Waiting for your feedback.

Thanks,

**Comment by jouino16** [comment]
Hello @marwachelly,

Thank you for your quick reply.

It is not a bug that I encountered, but it is a potential one. So I don't have neither a reproduction steps nor a screenshot.

It is technical, and should be checked by a developer. All he has to do is to read the code snippets I mentioned in the issue description and check if the code was written in purpose or not.

Thank you and have a great day!

**Comment by marwachelly** [comment]
Hello @jouino16 ,

So your suggession is to improve the code. 
@PrestaShop/prestashop-core-developers Do you think this code may result a bug?

Many thanks,

**Comment by jouino16** [comment]
Hello @marwachelly, yes exactly, I think that may result in an unwanted behaviour.

As I explained in the issue description: while implementing product hooks in a module and reading the code, I noticed a call to both save and update hooks in the `Product::update`. 

What I want (and what is expected) is to execute the save hook in the product add method and the update hook in the product update method. Currently, if I implement both save and update hooks and call the product update method, both hooks will be called, instead of calling only the update hook.

So I avoided implementing those 2 hooks, and I have implemented `hookActionObjectProductAddAfter` and `hookActionObjectProductUpdateAfter` instead, as I want my hook to be called once.

**Comment by atomiix** [comment]
@jouino16 I don't think this is a bug, when you call `product::save()` either `product::update()` or `product::add()` is called but both methods never seems to be called in a row thus `actionProductSave` hook is called only once per `product::update()/product::save()` while `actionProductUpdate` hook is called only when `product::update()` is called.

Am I missing something?

**Comment by jouino16** [comment]
@atomiix 

> I don't think this is a bug, when you call product::save() either product::update() or product::add() is called but both methods never seems to be called in a row

⇒ I understand what you are referring to this method `ObjectModel::save()` (correct me if I'm wrong)
```
public function save($null_values = false, $auto_date = true)
{
    return (int) $this->id > 0 ? $this->update($null_values) : $this->add($auto_date, $null_values);
}
```

> thus actionProductSave hook is called only once per product::update()/product::save() while actionProductUpdate hook is called only when product::update() is called.

⇒ Yes, it is true, but why would I trigger a `actionProductSave` hook when I call `product::update()`?
In my humble opinion, a call to `product::add()` should trigger the `actionProductSave` hook and a call to `product::update()` should trigger the `actionProductUpdate` hook.

Let's have a look at the `Product::add()` method:
```
public function add($autodate = true, $null_values = false)
{
    if ($this->is_virtual) {
        $this->product_type = ProductType::TYPE_VIRTUAL;
    }

    if (!parent::add($autodate, $null_values)) {
        return false;
    }

    $id_shop_list = Shop::getContextListShopID();
    if ($this->getType() == Product::PTYPE_VIRTUAL) {
        foreach ($id_shop_list as $value) {
            StockAvailable::setProductOutOfStock((int) $this->id, OutOfStockType::OUT_OF_STOCK_AVAILABLE, $value);
        }

        if ($this->active && !Configuration::get('PS_VIRTUAL_PROD_FEATURE_ACTIVE')) {
            Configuration::updateGlobalValue('PS_VIRTUAL_PROD_FEATURE_ACTIVE', '1');
        }
    } else {
        foreach ($id_shop_list as $value) {
            StockAvailable::setProductOutOfStock((int) $this->id, OutOfStockType::OUT_OF_STOCK_DEFAULT, $value);
        }
    }

    $this->setGroupReduction();
    Hook::exec('actionProductSave', ['id_product' => (int) $this->id, 'product' => $this]);

    return true;
}
```
⇒ You can see the call to the `actionProductSave` hook through the following statement. And this is OK.
`Hook::exec('actionProductSave', ['id_product' => (int) $this->id, 'product' => $this]);`

Now let's have a look at the `Product::update()` method:
```
public function update($null_values = false)
{
    if ($this->is_virtual) {
        $this->product_type = ProductType::TYPE_VIRTUAL;
    }

    $return = parent::update($null_values);
    $this->setGroupReduction();

    // Sync stock Reference, EAN13, MPN and UPC
    if (Configuration::get('PS_ADVANCED_STOCK_MANAGEMENT') && StockAvailable::dependsOnStock($this->id, Context::getContext()->shop->id)) {
        Db::getInstance()->update('stock', [
            'reference' => pSQL($this->reference),
            'ean13' => pSQL($this->ean13),
            'isbn' => pSQL($this->isbn),
            'upc' => pSQL($this->upc),
            'mpn' => pSQL($this->mpn),
        ], 'id_product = ' . (int) $this->id . ' AND id_product_attribute = 0');
    }

    Hook::exec('actionProductSave', ['id_product' => (int) $this->id, 'product' => $this]);
    Hook::exec('actionProductUpdate', ['id_product' => (int) $this->id, 'product' => $this]);
    if ($this->getType() == Product::PTYPE_VIRTUAL && $this->active && !Configuration::get('PS_VIRTUAL_PROD_FEATURE_ACTIVE')) {
        Configuration::updateGlobalValue('PS_VIRTUAL_PROD_FEATURE_ACTIVE', '1');
    }

    return $return;
}
```
⇒ You can see both calls to the `actionProductSave` and `actionProductUpdate` hooks through the following statements.
```
Hook::exec('actionProductSave', ['id_product' => (int) $this->id, 'product' => $this]);
Hook::exec('actionProductUpdate', ['id_product' => (int) $this->id, 'product' => $this]);
```
And here is the question: why there is a call to the `actionProductSave` hook after a product update call?

If my goal is to implement a separate hook for each addition or update event, it wouldn't be possible with these 2 hooks `actionProductSave` and `actionProductUpdate`.

**Comment by atomiix** [comment]
@jouino16 I think `save` is different from `add`. Let's say you're editing a text document, at the end you'll `save` it. You updated the content but the action you made was `save`. I think it's something similar here. As a developper I think it's nice to be able to subscribe to `actionProductSave` so each time a product is created/modified you can get notified with only one hook but also you can choose to subscribe to `actionProductUpdate` when you want to get notified **only** when the product is modified.

**Comment by jouino16** [comment]
@atomiix I understand your point of view.
So to which hook I should subscribe if I want to get notified **only** when a product is created (add event)?
This is the issue that I'm trying to point out. As I was in need of subscribing to a hook for **only update event** and another separate one for **only add event.**

**Comment by atomiix** [comment]
@jouino16 there's no easy way but I think you can use `actionObjectProductAddBefore` to see if there's an id in the object or not. If there's no id, it means it's a creation.

**Comment by jouino16** [comment]
@atomiix I should maybe give it a try.
Also, there is this hook "actionProductAdd" that is executed in both methods `AdminProductsController::processDuplicate` and `AdminProductsController::processAdd`.

My purpose behind opening this issue, is to check if there is a potential bug.

**Comment by kpodemski** [comment]
Since it's just a potential bug and this hasn't been touched for a long time, I'm closing this issue. Feel free to reopen if needed, and thank you for all the insights!
